master/internal/rm/kubernetesrm/pods.go

```diff
@@ -334,6 +364,185 @@ func (p *pods) getSystemResourceRequests(ctx *actor.Context) error {
+	listOptions := metaV1.ListOptions{
+		LabelSelector: fmt.Sprintf("%s=%s", determinedLabel, msg.allocationID),
+	}
+
+	pods, err := p.podInterface.List(context.TODO(), listOptions)
+	if err != nil {
+		return errors.Wrap(err, "error listing pods checking if they can be restored")
+	}
+
+	configMaps, err := p.configMapInterface.List(context.TODO(), listOptions)
+	if err != nil {
+		return errors.Wrap(err, "error listing config maps checking if they can be restored")
+	}
+	existingConfigMaps := make(map[string]bool)
+	for _, cm := range configMaps.Items {
+		existingConfigMaps[cm.Name] = true
+	}
+
+	var containerIDs []string
+	var k8sPods []*k8sV1.Pod
+	for _, pod := range pods.Items {
+		found := false
+		for _, container := range pod.Spec.Containers {
+			for _, env := range container.Env {
+				if env.Name == "DET_CONTAINER_ID" {
+					if !existingConfigMaps[pod.Name] {
+						p.deleteAllocationResources(ctx, pods, configMaps)
+						ctx.Respond(fmt.Errorf("pod missing config map %s", pod.Name))
+						return nil
+					}
+
+					p := pod
+					k8sPods = append(k8sPods, &p)
+					containerIDs = append(containerIDs, env.Value)
+
+					found = true
+					break
+				}
+			}
+			if found {
+				break
+			}
+		}
+	}
+
+	if len(k8sPods) != msg.numPods {
+		p.deleteAllocationResources(ctx, pods, configMaps)
+		ctx.Respond(fmt.Errorf("not enough pods found for allocation"))
+		return nil
+	}
+
+	var restoreResponses []reattachPodResponse
+	for i, containerID := range containerIDs {
+		resp, err := p.reattachPod(ctx, msg.taskActor, containerID,
+			k8sPods[i], msg.proxyPort, msg.slots, msg.logContext)
+		if err != nil {
+			p.deleteAllocationResources(ctx, pods, configMaps)
+			ctx.Respond(errors.Wrapf(err,
+				"error restoring pod with containerID %s", containerID))
+			return nil
+		}
+		restoreResponses = append(restoreResponses, resp)
+	}
+
+	ctx.Respond(restoreResponses)
+	return nil
+}
+
+func (p *pods) reattachPod(
+	ctx *actor.Context,
+	taskActor *actor.Ref,
+	containerID string,
+	pod *k8sV1.Pod,
+	proxyPort *sproto.ProxyPortConfig,
+	slots int,
+	logContext logger.Context,
+) (reattachPodResponse, error) {
+	startMsg := StartTaskPod{
+		TaskActor: taskActor,
+		Spec: tasks.TaskSpec{
+			ContainerID: containerID,
+		},
+		Slots:      slots,
+		LogContext: logContext,
+	}
+
+	newPodHandler := newPod(
+		startMsg,
+		p.cluster,
+		startMsg.Spec.ClusterID,
+		p.clientSet,
+		p.namespace,
+		p.masterIP,
+		p.masterPort,
+		p.masterTLSConfig,
+		p.loggingTLSConfig,
+		p.loggingConfig,
+		p.podInterface,
+		p.configMapInterface,
+		p.resourceRequestQueue,
+		p.leaveKubernetesResources,
+		p.slotType,
+		p.slotResourceRequests,
+		p.scheduler,
+		p.fluentConfig,
+	)
+
+	newPodHandler.restore = true
+	newPodHandler.logCtx["pod"] = pod.Name
+	newPodHandler.podName = pod.Name
+	newPodHandler.configMapName = pod.Name
+
+	if proxyPort != nil {
+		// TODO in pod.go cproto.Running p.ports is an array.
```

Comment:
one for each port exposed by env.ports in the job conf. one is added in `task_trial.go:43` for rendezvous, too. we can chat offline about this maybe?

Reply:
rendezvous and proxy port are the only ones that are used

Reply:
Talked offline, ended up going with setting the informational field ```ports``` on the container spec then reading that while restoring pods